File: node/src/transaction.ts
/*
 * Copyright 2020 IBM All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { CallOptions } from '@grpc/grpc-js';
import { common, gateway } from '@hyperledger/fabric-protos';
import { GatewayClient } from './client';
import { assertDefined } from './gateway';
import { Signable } from './signable';
import { SigningIdentity } from './signingidentity';
import { SubmittedTransaction, SubmittedTransactionImpl } from './submittedtransaction';
import { parseTransactionEnvelope } from './transactionparser';

/**
 * Represents an endorsed transaction that can be submitted to the orderer for commit to the ledger.
 */
export interface Transaction extends Signable {
    /**
     * Get the transaction result. This is obtained during the endorsement process when the transaction proposal is
     * run on endorsing peers.
     */
    getResult(): Uint8Array;

    /**
     * Get the transaction ID.
     */
    getTransactionId(): string;

    /**
     * Submit the transaction to the orderer to be committed to the ledger.
     * @param options - gRPC call options.
     * @throws {@link SubmitError}
     * Thrown if the gRPC service invocation fails.
     */
    submit(options?: CallOptions): Promise<SubmittedTransaction>;
}

export interface TransactionImplOptions {
    client: GatewayClient;
    signingIdentity: SigningIdentity;
    preparedTransaction: gateway.PreparedTransaction;
}

export class TransactionImpl implements Transaction {
    readonly #client: GatewayClient;
    readonly #signingIdentity: SigningIdentity;
    readonly #channelName: string;
    readonly #preparedTransaction: gateway.PreparedTransaction;
    readonly #envelope: common.Envelope;
    readonly #result: Uint8Array;

    constructor(options: Readonly<TransactionImplOptions>) {
        this.#client = options.client;
        this.#signingIdentity = options.signingIdentity;
        this.#preparedTransaction = options.preparedTransaction;

        const envelope = assertDefined(options.preparedTransaction.getEnvelope(), 'Missing envelope');
        this.#envelope = envelope;

        const { channelName, result } = parseTransactionEnvelope(envelope);
        this.#channelName = channelName;
        this.#result = result;
    }

    getBytes(): Uint8Array {
        return this.#preparedTransaction.serializeBinary();
    }

    getDigest(): Uint8Array {
        const payload = this.#envelope.getPayload_asU8();
        return this.#signingIdentity.hash(payload);
    }

    getResult(): Uint8Array {
        return this.#result;
    }

    getTransactionId(): string {
        return this.#preparedTransaction.getTransactionId();
    }

    async submit(options?: Readonly<CallOptions>): Promise<SubmittedTransaction> {
        await this.#sign();
        await this.#client.submit(this.#newSubmitRequest(), options);

        return new SubmittedTransactionImpl({
            client: this.#client,
            signingIdentity: this.#signingIdentity,
            transactionId: this.getTransactionId(),
            signedRequest: this.#newSignedCommitStatusRequest(),
            result: this.getResult(),
        });
    }

    setSignature(signature: Uint8Array): void {
        this.#envelope.setSignature(signature);
    }

    async #sign(): Promise<void> {
        if (this.#isSigned()) {
            return;
        }

        const signature = await this.#signingIdentity.sign(this.getDigest());
        this.setSignature(signature);
    }

    #isSigned(): boolean {
        const signatureLength = this.#envelope.getSignature_asU8()?.length || 0;
        return signatureLength > 0;
    }

    #newSubmitRequest(): gateway.SubmitRequest {
        const result = new gateway.SubmitRequest();
        result.setTransactionId(this.getTransactionId());
        result.setChannelId(this.#channelName);
        result.setPreparedTransaction(this.#envelope);
        return result;
    }

    #newSignedCommitStatusRequest(): gateway.SignedCommitStatusRequest {
        const result = new gateway.SignedCommitStatusRequest();
        result.setRequest(this.#newCommitStatusRequest().serializeBinary());
        return result;
    }

    #newCommitStatusRequest(): gateway.CommitStatusRequest {
        const result = new gateway.CommitStatusRequest();
        result.setChannelId(this.#channelName);
        result.setTransactionId(this.getTransactionId());
        result.setIdentity(this.#signingIdentity.getCreator());
        return result;
    }
}

File: pkg/client/transaction.go
/*
Copyright 2020 IBM All Rights Reserved.

SPDX-License-Identifier: Apache-2.0
*/

package client

import (
	"context"
	"fmt"
	"time"

	"github.com/golang/protobuf/proto"
	"github.com/hyperledger/fabric-protos-go/gateway"
)

// Transaction represents an endorsed transaction that can be submitted to the orderer for commit to the ledger.
type Transaction struct {
	client              gateway.GatewayClient
	signingID           *signingIdentity
	channelID           string
	preparedTransaction *gateway.PreparedTransaction
}

// Result of the proposed transaction invocation.
func (transaction *Transaction) Result() []byte {
	return transaction.preparedTransaction.Result.Payload
}

// Bytes of the serialized transaction.
func (transaction *Transaction) Bytes() ([]byte, error) {
	transactionBytes, err := proto.Marshal(transaction.preparedTransaction)
	if err != nil {
		return nil, fmt.Errorf("failed to marshall PreparedTransaction protobuf: %w", err)
	}

	return transactionBytes, nil
}

// Digest of the transaction. This is used to generate a digital signature.
func (transaction *Transaction) Digest() []byte {
	return transaction.signingID.Hash(transaction.preparedTransaction.Envelope.Payload)
}

// TransactionID of the transaction.
func (transaction *Transaction) TransactionID() string {
	return transaction.preparedTransaction.TransactionId
}

// Submit the transaction to the orderer for commit to the ledger.
func (transaction *Transaction) Submit() (*Commit, error) {
	if err := transaction.sign(); err != nil {
		return nil, err
	}

	// Build before the submit to avoid chance of errors after the submit
	statusRequest, err := transaction.newSignedCommitStatusRequest()
	if err != nil {
		return nil, err
	}

	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
	defer cancel()

	submitRequest := &gateway.SubmitRequest{
		TransactionId:       transaction.TransactionID(),
		ChannelId:           transaction.channelID,
		PreparedTransaction: transaction.preparedTransaction.Envelope,
	}
	_, err = transaction.client.Submit(ctx, submitRequest)
	if err != nil {
		return nil, fmt.Errorf("failed to submit transaction to the orderer: %w", err)
	}

	return newCommit(transaction.client, transaction.signingID, transaction.TransactionID(), statusRequest), nil
}

func (transaction *Transaction) sign() error {
	if transaction.isSigned() {
		return nil
	}

	digest := transaction.Digest()
	signature, err := transaction.signingID.Sign(digest)
	if err != nil {
		return err
	}

	transaction.setSignature(signature)

	return nil
}

func (transaction *Transaction) isSigned() bool {
	return len(transaction.preparedTransaction.Envelope.Signature) > 0
}

func (transaction *Transaction) setSignature(signature []byte) {
	transaction.preparedTransaction.Envelope.Signature = signature
}

func (transaction *Transaction) newSignedCommitStatusRequest() (*gateway.SignedCommitStatusRequest, error) {
	creator, err := transaction.signingID.Creator()
	if err != nil {
		return nil, fmt.Errorf("failed to serialize identity: %w", err)
	}

	request := &gateway.CommitStatusRequest{
		ChannelId:     transaction.channelID,
		TransactionId: transaction.TransactionID(),
		Identity:      creator,
	}

	requestBytes, err := proto.Marshal(request)
	if err != nil {
		return nil, err
	}

	signedRequest := &gateway.SignedCommitStatusRequest{
		Request: requestBytes,
	}
	return signedRequest, nil
}
